File: Programming - 2/Lab_5_6/Lab_5_CPP/Back/CustomDigitString.cpp
#include <cctype>
#include "CustomDigitString.h"

CustomDigitString::CustomDigitString() : CustomString::CustomString()
{ }

CustomDigitString::CustomDigitString(char* line)
{
	SetLine(line);
}

void CustomDigitString::SetLine(char* line)
{
	if (Line != nullptr)
		delete[] Line;
	Length = 0;
	Line = new char[1];
	int j = 0;
	while (line[j] != '\0')
	{
		if (isdigit(line[j]))
		{
			Length++;
			char* tempLine = new char[Length + 1];
			for (int i = 0; i < Length - 1; i++)
			{
				tempLine[i] = Line[i];
			}
			tempLine[Length - 1] = line[j];
			delete[] Line;
			Line = tempLine;
		}
		j++;
	}
	Line[Length] = '\0';
}

void CustomDigitString::Flip()
{
	char* tempLine = new char[Length + 1];
	for (int i = 0; i < Length; i++)
	{
		tempLine[i] = Line[Length - 1 - i];
	}
	tempLine[Length] = '\0';
	delete[] Line;
	Line = tempLine;
}

File: Programming - 2/Lab_5_6/Lines/CustomDigitString.cs
using System;

namespace Lines
{
    public class CustomDigitString : CustomString
    {
        public CustomDigitString()
        { }
        public CustomDigitString(char[] line)
        {
            Line = line;
        }
        /// <summary>
        /// Setter only adds digits to the line
        /// </summary>
        public new char[] Line
        {
            get => base.Line;


            set
            {
                char[] tempLine = new char[value.Length];
                int nOfNonDigits = 0;
                for (int i = 0, iTemp = 0; i < value.Length; i++, iTemp++)
                {
                    if (!Char.IsDigit(value[i]))
                    {
                        nOfNonDigits++;
                        iTemp--;
                        continue;
                    }
                    tempLine[iTemp] = value[i];
                }

                line = new char[value.Length - nOfNonDigits];
                for (int i = 0; i < line.Length; i++)
                {
                    line[i] = tempLine[i];
                }
            }
        }
        public void Flip()
        {
            char[] tempLine = new char[Length];
            for (int i = 0; i < Length; i++)
            {
                tempLine[i] = Line[Length - 1 - i];
            }
            Line = tempLine;
        }
    }
}
